src/Npgsql/TaskTimeoutAndCancellation.cs

```diff
@@ -0,0 +1,132 @@
+        Task<TResult>? task = default;
+        await WaitAsync(ct => (Task)(task = getTaskFunc(ct)), timeout, cancellationToken, useLegacyImplementation);
+        return await task!;
+    }
+
+    /// <summary>
+    /// Utility that executes a non-cancellable task while allowing to timeout and/or cancel awaiting for it.
+    /// If the given task does not complete within <paramref name="timeout"/>, a <see cref="TimeoutException"/> is thrown.
+    /// </summary>
+    /// <param name="getTaskFunc">Gets a <see cref="Task"/>.</param>
+    /// <param name="timeout">The timeout after which the <see cref="Task"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
+    /// <param name="useLegacyImplementation">An optional <see cref="Boolean"/> parameter for testing only. If set to true the method uses pre-dotnet 6.0 implementation for cancellation and timeout enforcement.</param>
+    /// <returns>The <see cref="Task"/> representing the asynchronous wait.</returns>
+    internal static async Task WaitAsync(Func<CancellationToken, Task> getTaskFunc, NpgsqlTimeout timeout, CancellationToken cancellationToken, bool useLegacyImplementation = false)
+    {
+        using var combinedCts = timeout.IsSet ? CancellationTokenSource.CreateLinkedTokenSource(cancellationToken) : null;
+        var task = getTaskFunc(combinedCts?.Token ?? cancellationToken);
+        try
+        {
+            try
+            {
+                await WithCancellationAndTimeoutLocal(task);
+            }
+            catch (TimeoutException) when (!task!.IsCompleted)
+            {
+                // Attempt to stop the Task in progress.
+                combinedCts?.Cancel();
+                throw;
+            }
+        }
+        catch
+        {
+            // Prevent unobserved Task notifications by observing the failed Task exception.
+            // To test: comment the next line out and re-run TaskExtensionsTest.DelayedFaultedTaskCancellation.
+            _ = task.ContinueWith( t => _ = t.Exception, CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Current);
+            throw;
+        }
+
+        Task WithCancellationAndTimeoutLocal(Task t)
+        {
+            if (useLegacyImplementation)
+            {
+                return WithCancellationAndTimeoutImplLegacy(t, timeout, cancellationToken);
+            }
+            return WithCancellationAndTimeout(t, timeout, cancellationToken);
+        }
+    }
+    
+#if NET6_0_OR_GREATER
+    static Task WithCancellationAndTimeoutImpl(Task task, NpgsqlTimeout timeout, CancellationToken cancellationToken)
+        => task.WaitAsync(timeout.CheckAndGetTimeLeft(), cancellationToken);
+#endif
+
+    static async Task WithCancellationAndTimeoutImplLegacy(Task task, NpgsqlTimeout timeout, CancellationToken cancellationToken)
```

Comment:
As written elsewhere, I think this code should be #ifdef'd to pre-.NET 6.0.

The simplest seems to be to copy the implementation of the .NET 6.0 Task.WaitAsync into an extension method in our project. This would allow the code here to simply use WaitAsync without caring whether it's coming from .NET 6.0 or from our own copy. We've done this for various other APIs which were added in later TFMs, see the "shims" directory for some examples (this is where I'd put it).

Reply:
Let's agree that the class's public interface is agnostic to whether it is .NET 6.0 or an earlier runtime.
It is not an extension class. The only reason that, as explained earlier, this class provides a clear way of canceling the Task that is about to become abandoned. 
We are not re- implementing NET 6.0 Task.WaitAsync extension API. That API is there but is easy to misuse BTW. It carries a risk of creating the same complex problem as we are trying to fix here: abandoning the original Task.

Reply:
I think in my comment above I confused the higher-level task of preventing the unobserved exception and the lower-level task of implementing the cancellation and timeout - you're right that WaitAsync does the former but not the latter.

But leaving aside the former and concentrating on the factoring of the latter... The current code as it is has TaskTimeoutAndCancellation.WithCancellationAndTimeout, which simply wraps Task.WaitAsync for .NET6.0, and WithCancellationAndTimeout, which provides the identical functionality for older TFMs (and with the parameters etc.). In other words, WithCancellationAndTimeout seems to provide the exact same functionality that .NET 6.0 WaitAsync does - it really is a backwards-compat simple shim. So unless I'm mistaken, there's nothing preventing WithCancellationAndTimeout from being implemented as a Task.WaitAsync extension method, providing the exact same API as the .NET 6.0 but for earlier TFMs, no? This would simplify things and move the legacy code out to the shims directory.

Reply:
I think I understand. Will make a change and see if that's what you want.

Reply:
Looks great, thanks! Could you please move the legacy WaitAsync implementation to a TaskExtensions.cs under the Shims directory? That's how we usually take care of this kind of thing.

Reply:
Will do.

Reply:
Pushed the update. Please run the checks.

Reply:
One more thing: do we need the copyright notices for each new module? There weren't any in the former TaskExtensions.cs and I kept it that way.